commons-vfs2-sandbox/src/main/java/org/apache/commons/vfs2/provider/smb2/Smb2ClientWrapper.java

```diff
@@ -0,0 +1,234 @@
+            } else {
+                userName = userString;
+            }
+        }
+
+        String password = rootName.getPassword();
+        String authDomain = (userString.contains(";") ?
+                userString.substring(0, userString.indexOf(";")) : null);
+
+        //if username == "" the client tries to authenticate "anonymously". It's also possible to submit "guest" as username
+        AuthenticationContext authContext = new AuthenticationContext(userName, password.toCharArray(), authDomain);
+
+        //a connection stack is: SMBClient > Connection > Session > DiskShare
+        try {
+            connection = smbClient.connect(rootName.getHostName());
+            session = connection.authenticate(authContext);
+            String share = ((Smb2FileName) rootName).getShareName();
+            diskShare = (DiskShare) session.connectShare(share);
+        } catch (Exception e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Error while creating connection to " + rootName.getHostName());
+            }
+            throw new FileSystemException("vfs.provider.smb2/connect.error", rootName.getHostName(), e);
+        }
+    }
+
+    public GenericFileName getRoot() {
+
+        return root;
+    }
+
+    /**
+     * Returns file information of a given file.
+     *
+     * @param relativePath of file
+     * @return FileAllInformation object
+     */
+    public FileAllInformation getFileInfo(String relativePath) {
+
+        try {
+            return diskShare.getFileInformation(relativePath);
+        } catch (Exception e) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Could not get information for file: " + relativePath);
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Returns a write handle on the file.
+     *
+     * @param path path to file
+     * @return DiskEntry file
+     */
+    public DiskEntry getDiskEntryWrite(String path) {
+
+        return diskShare.open(path,
+                EnumSet.of(AccessMask.MAXIMUM_ALLOWED),
+                EnumSet.of(FileAttributes.FILE_ATTRIBUTE_NORMAL),
+                EnumSet.of(SMB2ShareAccess.FILE_SHARE_WRITE),
+                SMB2CreateDisposition.FILE_OPEN_IF,
+                EnumSet.of(SMB2CreateOptions.FILE_NO_COMPRESSION));
+    }
+
+    /**
+     * Creates a folder on a given path.
+     *
+     * @param path of folder
+     */
+    public void createFolder(String path) {
+
+        DiskEntry de = getDiskEntryFolderWrite(path);
+        de.close();
+    }
+
+    /**
+     * Returns a handle for a folder.
+     *
+     * @param path folder path
+     * @return DiskEntry
+     */
+    public DiskEntry getDiskEntryFolderWrite(String path) {
+
+        return diskShare.openDirectory(path,
+                EnumSet.of(AccessMask.GENERIC_ALL),
+                EnumSet.of(FileAttributes.FILE_ATTRIBUTE_NORMAL),
+                EnumSet.of(SMB2ShareAccess.FILE_SHARE_READ),
```

Comment:
@sachithKay - can you please explain why write folder receive SMB2ShareAccess.FILE_SHARE_READ access and not SMB2ShareAccess.FILE_SHARE_WRITE ?